### PR TITLE
Restore DIGEST-MD5 auth support

### PR DESCRIPTION
**Pull request by ashafer01**
Restoring the working auth QOP implementation for DIGEST-MD5, and cleaning up non-working/incomplete auth-int support. This aligns with the level of support for other mechs (other than GSSAPI).

DIGEST-MD5 has been tested against an LDAP server.

### CODE REVIEW

**Comment by thobbs** [comment]
Awesome, I'm very excited about having working DIGEST-MD5 support!  Thank you for your work so far.  I'll do a quick review, and then we just need to sort out the tests that are failing in Travis.

File: puresasl/mechanisms.py [hunk 1]
```diff
@@ -52,6 +53,8 @@ class Mechanism(object):
 
     def __init__(self, sasl, **props):
         self.sasl = sasl
+        self.qops = [QOP.AUTH]
+        self.qop = QOP.AUTH
```
Comment (thobbs) [inline puresasl/mechanisms.py]: Can you clarify why this change is needed instead of using the QOPs from `sasl`?  That would be a breaking change from the current API (and we would need to remove the `qops` param from `SaslClient` if this is done).

File: puresasl/mechanisms.py [hunk 2]
```diff
@@ -103,7 +106,7 @@ def _pick_qop(self, server_qop_set):
         Choose a quality of protection based on the user's requirements and
         what the server supports.
         """
-        configured_qops = set(_b(qop) if isinstance(qop, str) else qop for qop in self.sasl.qops)  # normalize user-defined config
+        configured_qops = set(_b(qop) if isinstance(qop, str) else qop for qop in self.qops)  # normalize user-defined config
```
Comment (thobbs) [inline puresasl/mechanisms.py]: Same question as above

File: puresasl/mechanisms.py [hunk 3]
```diff
@@ -187,6 +190,56 @@ def process(self, challenge=None):
     def dispose(self):
         self.password = None
 
+    def wrap(self, outgoing):
+        if self.qop == 'auth-int' or self.qop == 'auth-conf':
+            raise Exception('{0} QoP not supported for CRAM-MD5'.format(self.qop))
```
Comment (thobbs) [inline puresasl/mechanisms.py]: If it's primarily to add a notion of supported QOPs to mechanisms (which seems to be the case here), we could just add a hard-coded list to each mechanism and check that up front.

File: puresasl/mechanisms.py [hunk 4]
```diff
@@ -187,6 +190,56 @@ def process(self, challenge=None):
...
+    """
+    Enclose in quotes and escape internal slashes and double quotes.
+
+    :param text: A Unicode or byte string.
+    """
+    text = bytes(text)
+    return b'"' + text.replace(b'\\', b'\\\\').replace(b'"', b'\\"') + b'"'
+
 
 # TODO: incomplete, not tested
```
Comment (thobbs) [inline puresasl/mechanisms.py]: I suppose we want to remove this :)

File: puresasl/mechanisms.py (hunk 1 above)
Comment (ashafer01) [inline puresasl/mechanisms.py]: The intention, as you say below, is to add supported QOPs to mechanisms. I did fail to take into account client-selected QOPs when making this change - I'll get that fixed

File: puresasl/mechanisms.py (hunk 3 above)
Comment (ashafer01) [inline puresasl/mechanisms.py]: Yes, you're right, it should be checked up front. Checking in wrap/unwrap should only be necessary if supporting multiple QOPs.

**Comment by ashafer01** [comment]
@thobbs tests are passing now! Please do a final review at your convenience. Thanks! :D

**Comment by thobbs** [comment]
I made some minor changes in another branch (mostly avoiding the new dependency on ``six``) and merged everything in.  Thanks again!

**Comment by thobbs** [comment]
Oh, and I'll go ahead and release a new version.
